src/PushMod_MesaUpdate/Plugin.cs

```diff
@@ -1,159 +1,390 @@
+            currentCharge += Time.deltaTime;
+            currentCharge = Mathf.Clamp(currentCharge, 0f, MAX_CHARGE);
+        }
+    }
+
+    /// <summary>
+    /// Handles input for charging and releasing the push action.
+    /// Charging begins on key down and applies the push on key up.
+    /// </summary>
+    private void HandleChargeInput() {
+        if (Plugin.PConfig.CanCharge) {
+            if ((Input.GetKeyDown(Plugin.PConfig.PushKey) || Input.GetKeyDown(Plugin.PConfig.SelfPushKey)) && !isCharging && coolDownLeft <= 0f) {
+                isCharging = true;
+                currentCharge = 0f;
+                Plugin.Log.LogInfo("Started charging push...");
+            }
+            if (Input.GetKeyUp(Plugin.PConfig.PushKey) && !Input.GetKey(Plugin.PConfig.SelfPushKey) && isCharging) {
+                isCharging = false;
+                TryPushTarget();
+            }
+            if (Input.GetKeyUp(Plugin.PConfig.SelfPushKey) && !Input.GetKey(Plugin.PConfig.PushKey) && isCharging) {
+                isCharging = false;
+                TryPushSelf();
+            }
+        }
+        else {
+            if (Input.GetKeyDown(Plugin.PConfig.PushKey) && coolDownLeft <= 0f) {
+                TryPushTarget();
+            }
+            if (Input.GetKeyDown(Plugin.PConfig.SelfPushKey) && coolDownLeft <= 0f) {
+                TryPushSelf();
             }
-            Plugin.Log.LogInfo("Sending Push RPC Event");
-            localCharacter.view.RPC("PushPlayer_Rpc", RpcTarget.All, new object[]
-            {
-                pushedCharacter.view.ViewID,
-                forceDirection,
-                localCharacter.view.ViewID
-            });
         }
     }
 
-    private Character GetCharacter(GameObject obj) {
-        if (!obj.TryGetComponent(out Character character)) {
-            if (obj.transform.parent == null) return character;
-            return GetCharacter(obj.transform.parent.gameObject);
+    /// <summary>
+    /// Attempts to perform a push using a forward raycast from the main camera.
+    /// If a valid character is hit, applies force via RPC.
+    /// </summary>
+    private void TryPushTarget() {
+        if (mainCamera is null) return;
+
+        // Perform raycast from camera forward within push range
+        if (!Physics.Raycast(mainCamera.transform.position, mainCamera.transform.forward, out RaycastHit hitInfo, PUSH_RANGE, LayerMask.GetMask("Character")))
+            return;
+
+        // Retrieve the Character component from the hit object
+        if (GetCharacter(hitInfo.transform.gameObject) is not Character pushedCharacter) return;
+        if (pushedCharacter == localCharacter) return;
+
+        // Calculate final push force with multipliers
+        float chargeMultiplier = 1f + (currentCharge / MAX_CHARGE) * CHARGE_FORCE_MULTIPLIER;
+        float bingBongMultiplier = bingBong ? BINGBONG_MULTIPLIER : 1f;
+        float totalMultiplier = bingBongMultiplier + chargeMultiplier;
```

Comment:
I think this line may be wrong given the bing-bong multiplier is 1 when absent. Maybe change it to this?

```suggestion
        float totalMultiplier = bingBongMultiplier * chargeMultiplier;
```